core/src/main/java/com/transferwise/common/gracefulshutdown/strategies/TaskSchedulersGracefulShutdownStrategy.java

```diff
@@ -1,86 +1,52 @@
 package com.transferwise.common.gracefulshutdown.strategies;
 
 import com.transferwise.common.gracefulshutdown.GracefulShutdownStrategy;
+import com.transferwise.common.gracefulshutdown.utils.ShutdownUtils;
 import java.util.ArrayList;
 import java.util.HashSet;
 import java.util.List;
-import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executor;
 import java.util.concurrent.Executors;
-import java.util.concurrent.ScheduledThreadPoolExecutor;
 import java.util.concurrent.atomic.AtomicInteger;
+import lombok.AllArgsConstructor;
+import lombok.Builder;
+import lombok.NoArgsConstructor;
 import lombok.extern.slf4j.Slf4j;
 import org.springframework.beans.factory.annotation.Autowired;
 import org.springframework.context.ApplicationContext;
 import org.springframework.scheduling.TaskScheduler;
-import org.springframework.scheduling.concurrent.ConcurrentTaskScheduler;
-import org.springframework.scheduling.concurrent.ThreadPoolTaskScheduler;
 
 @Slf4j
+@NoArgsConstructor
+@AllArgsConstructor
+@Builder(toBuilder = true)
 public class TaskSchedulersGracefulShutdownStrategy implements GracefulShutdownStrategy {
 
   @Autowired
   private ApplicationContext applicationContext;
 
-  private List<TaskScheduler> taskSchedulers = new ArrayList<>();
+  private final List<TaskScheduler> taskSchedulers = new ArrayList<>();
 
-  private AtomicInteger inProgressShutdowns = new AtomicInteger();
+  private final AtomicInteger inProgressShutdowns = new AtomicInteger();
```

Comment:
Right now we decrement this as soon as call a `shutdown()` on the executor. But already submitted tasks may still work in other threads as `shutdown()` is not blocking and not waiting for actual termination. So we will return true on `canShutdown()` as soon as call shutdown on all executors. However already started tasks may still be running. Is it a correct behaviour?

Reply:
`Is it a correct behaviour?`
No, this is a bug and should be fixed.

Reply:
About the `Things to keep in mind` section:
Basically as I understand it it's about adding more steps to the shutdown process. Currently we have the `prepareForShutdown` and `applicationTerminating` lifecycle hooks. Technically you could use that to do the same thing. For example use the `applicationTerminating` as a way to call `shutdownNow()` if needed with additional timeout.

Reply:
I was thinking about this, but it will require to make `applicationTerminating ` blocking and additionally run shutdown logic there which may be a little bit misleading given name of `applicationTerminating` + the fact, that strategy method `canShutdown` would return true, but it is not the case. For external developer like me this step `applicationTerminating` seems more like additional cleanup step after shutdown, than a phase for shutdown itself.

Reply:
As discussed, we reverted the `2 * shutdownTimeout` logic as it doesn't make sense to make it global if nothing else is being triggered when first `shutdownTimeout` is passed.